### DISCUSSION TITLE
Feature request: Run or operating hours counter

### DISCUSSION THREAD

**Discussion by ajaypala** [arendst/Tasmota#17355] [Ideas]
It would be good know how many hours a device/load/appliance has been in operation. Similar to the Energy Today, Yesterday and Total, we could have hours run/operating Today, Yesterday and Total. This was mentioned previously in https://github.com/arendst/Tasmota/issues/4019

The operating status can be derived from the relay status or if the device has power monitoring, when the power is greater than some threshold. 

Typical use cases would be for a pump or generator or other machine that needs maintenance after so many hours run. For device with power monitoring, it would be good to know how many hours it was actually working (eg a fridge or air conditioner or water heater).

**Comment by barbudor** [comment]
Not all load behave the same: power threshold and timeout to consider the appliance is on or off would be different from one to another
I'm doing that for TV, washing machine and a few other stuff using a combination of rules and influxdb. rules are adapted for each device

For example I have a pump in the basement powered through a plug which relay is always on. The pump automatically start when the water level is above a certain value. The plug transmit OPERATING ON and OPERATING OFF status based on those rules:
```
Rule1 
  on Energy#Power>%Mem1% do if (Var1==0) Publish stat/%topic%/OPERATING ON;Var1 1;Var2 60;Var4 ON endif endon
  on Energy#Power<=%Mem2% do RuleTimer1 %Mem3% endon on Energy#Power>%Mem2% do RuleTimer1 0 endon
Rule2 
  on Rules#Timer=1 do IF (Var1==1) Publish stat/%topic%/OPERATING OFF;Var1 0;Var2 %var3%;Var4 OFF endif endon
  on Time#Minute|%var2% do Publish tele/%topic%/OPERATING %Var4% endon
Rule1 1
Rule1 5
Rule2 1
```
Node-red forward to InfluxDB which calculate the number of working hours per month

**Reply by sfromis** [reply]
When using an energy monitoring device, I'm doing something similar backend, calculating the power draw, based on using PowerDelta to have reporting as things change. As some device can have quite marked spikes, I'm smoothening short spikes.

Instead of using Power (which is at the instant, not averaged), I'm using the difference between two Today numbers to calculate the energy consumption in each interval, which has proved to be a more reliable way of detecting what goes up and down, like when having more than one device behind the same energy monitoring, the jumps can allow me to see what device is going on/off.

Thus, I can have reports saying things like my kettle was on for 4 minutes, using 93 Wh of power, at the same time as I also have other devices on the same outlet.

For this, I need no rules to do extra publish on the device, but next step (with ESP32 devices) is going to be using Berry to do the same averaging stuff local on the energy monitoring/relay box, and have a small display updating the most interesting ones, without needing a corner on my desktop computer screen setup. I could also use the same Berry code to subscribe to the energy reporting from older ESP8266 devices without the ability to run Berry locally.

**Comment by ajaypala** [comment]
Thanks for the comments. I should point out, that I'm using the Tasmota devices standalone without a server (influxDB, Node-Red, etc). Each Tasmota device only has one appliance/load connected to keep it simple. 

My pump example is controlled by the Tasmota relay, so just need to count the number of hours the relay is on. Then I know after x hours of operating that I need to clean the strainer or perform some other maintenance.

I appreciate each appliance/load will require different power thresholds to enable or disable counting hours and most loads will have sufficiently different power readings between standby and operating. For my air conditioner, the power reading between the fan running and the fan with compressor is significant and reliable for this application.

**Reply by sfromis** [reply]
Tasmota is not designed for "standalone" operation, the "M" in the name is for MQTT, and Tasmota focuses on making information available, not on creating historical reports. You are making operation more complex without having a backend.

**Comment by barbudor** [comment]
You can still enhance tasmota using Scripting Language, or Berry scripting (the later being available only on ESP32)
People have made graphs and other things like that
A bit of learning curve
